File: celesta-test/score/sequences/testSequences.py
# coding=UTF-8

from ru.curs.celesta.unit import TestClass, CelestaTestCase
from ru.curs.celesta import CelestaException

from sequences._sequences_orm import s1Sequence, s2Sequence, s3Sequence, s4Sequence, s5Sequence, \
    s6Sequence, s7Sequence, s8Sequence, s9Sequence, s10Sequence, \
    s11Sequence, s12Sequence, s13Sequence, s14Sequence, s15Sequence, t1Cursor

@TestClass
class TestSequence(CelestaTestCase):

    def testS1(self):
        s = s1Sequence(self.context)

        self.assertEquals(1L, s.nextValue())
        self.assertEquals(2L, s.nextValue())

    def testS2(self):
        s = s2Sequence(self.context)

        self.assertEquals(3L, s.nextValue())
        self.assertEquals(4L, s.nextValue())

    def testS3(self):
        s = s3Sequence(self.context)

        self.assertEquals(1L, s.nextValue())
        self.assertEquals(3L, s.nextValue())

    def testS4(self):
        s = s4Sequence(self.context)

        self.assertEquals(1L, s.nextValue())
        self.assertEquals(-1L, s.nextValue())

    def testS5(self):
        s = s5Sequence(self.context)

        self.assertEquals(5L, s.nextValue())
        self.assertEquals(4L, s.nextValue())

    def testS6(self):
        s = s6Sequence(self.context)

        self.assertEquals(5L, s.nextValue())
        self.assertEquals(4L, s.nextValue())
        #end of sequence
        self.assertThrows(CelestaException, nextVal, s)

    def testS7(self):
        s = s7Sequence(self.context)

        self.assertEquals(-3L, s.nextValue())
        self.assertEquals(-5L, s.nextValue())
        #end of sequence
        self.assertThrows(CelestaException, nextVal, s)

    def testS8(self):
        s = s8Sequence(self.context)

        self.assertEquals(-1L, s.nextValue())
        #end of sequence
        self.assertThrows(CelestaException, nextVal, s)

    def testS9(self):
        s = s9Sequence(self.context)

        self.assertEquals(5L, s.nextValue())
        self.assertEquals(6L, s.nextValue())

    def testS10(self):
        s = s10Sequence(self.context)

        self.assertEquals(-5L, s.nextValue())
        self.assertEquals(-4L, s.nextValue())

    def testS11(self):
        s = s11Sequence(self.context)

        self.assertEquals(1L, s.nextValue())
        self.assertEquals(2L, s.nextValue())

    def testS12(self):
        s = s12Sequence(self.context)

        self.assertEquals(1L, s.nextValue())
        self.assertEquals(2L, s.nextValue())
        self.assertEquals(3L, s.nextValue())
        self.assertEquals(4L, s.nextValue())
        self.assertEquals(5L, s.nextValue())
        #new cycle
        self.assertEquals(1L, s.nextValue())

    def testS13(self):
        s = s13Sequence(self.context)

        self.assertEquals(1L, s.nextValue())
        self.assertEquals(0L, s.nextValue())
        self.assertEquals(-1L, s.nextValue())
        #new cycle
        self.assertEquals(1L, s.nextValue())

    def testS14AndS15(self):
        self._testS14OrS15(s14Sequence(self.context))
        self._testS14OrS15(s15Sequence(self.context))

    def testDefaultPkColumnValueWithSequence(self):
        c = t1Cursor(self.context)

        c.insert()
        c.last()
        self.assertEquals(4, c.id)
        c.clear()
        c.insert()
        c.last()
        self.assertEquals(6, c.id)


    def _testS14OrS15(self, s):
        lastValue = s.nextValue()
        self.assertEquals(5L, lastValue)
        lastValue = s.nextValue()
        self.assertEquals(7L, lastValue)

        while (lastValue < 55):
            lastValue = s.nextValue()

        #new cycle
        lastValue = s.nextValue()
        self.assertEquals(5L, lastValue)

def nextVal(s):
    s.nextValue()

File: celesta-test/src/test/java/ru/curs/celesta/script/TestSequences.java
package ru.curs.celesta.script;

import org.junit.jupiter.api.TestTemplate;
import ru.curs.celesta.CallContext;
import ru.curs.celesta.CelestaException;
import ru.curs.celesta.dbutils.Sequence;
import sequences.S10Sequence;
import sequences.S11Sequence;
import sequences.S12Sequence;
import sequences.S13Sequence;
import sequences.S14Sequence;
import sequences.S15Sequence;
import sequences.S1Sequence;
import sequences.S2Sequence;
import sequences.S3Sequence;
import sequences.S4Sequence;
import sequences.S5Sequence;
import sequences.S6Sequence;
import sequences.S7Sequence;
import sequences.S8Sequence;
import sequences.S9Sequence;
import sequences.T1Cursor;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertThrows;


class TestSequences implements ScriptTest {

    @TestTemplate
    void testS1(CallContext context) {
        S1Sequence s = new S1Sequence(context);

        assertEquals(1L, s.nextValue());
        assertEquals(2L, s.nextValue());
        assertEquals(3L, s.nextValue());
    }

    @TestTemplate
    void testS2(CallContext context) {
        S2Sequence s = new S2Sequence(context);

        assertEquals(3L, s.nextValue());
        assertEquals(4L, s.nextValue());
        assertEquals(5L, s.nextValue());
    }

    @TestTemplate
    void testS3(CallContext context) {
        S3Sequence s = new S3Sequence(context);

        assertEquals(1L, s.nextValue());
        assertEquals(3L, s.nextValue());
        assertEquals(5L, s.nextValue());
    }

    @TestTemplate
    void testS4(CallContext context) {
        S4Sequence s = new S4Sequence(context);

        assertEquals(1L, s.nextValue());
        assertEquals(-1L, s.nextValue());
        assertEquals(-3L, s.nextValue());
    }

    @TestTemplate
    void testS5(CallContext context) {
        S5Sequence s = new S5Sequence(context);

        assertEquals(5L, s.nextValue());
        assertEquals(4L, s.nextValue());
    }

    @TestTemplate
    void testS6(CallContext context) {
        S6Sequence s = new S6Sequence(context);

        assertEquals(5L, s.nextValue());
        assertEquals(4L, s.nextValue());

        //end of sequence
        assertThrows(CelestaException.class,
                s::nextValue);
    }

    @TestTemplate
    void testS7(CallContext context) {
        S7Sequence s = new S7Sequence(context);

        assertEquals(-3L, s.nextValue());
        assertEquals(-5L, s.nextValue());
        //end of sequence
        assertThrows(CelestaException.class,
                s::nextValue);
    }

    @TestTemplate
    void testS8(CallContext context) {
        S8Sequence s = new S8Sequence(context);
        assertEquals(-1L, s.nextValue());
        //end of sequence
        assertThrows(CelestaException.class,
                s::nextValue);

    }

    @TestTemplate
    void testS9(CallContext context) {
        S9Sequence s = new S9Sequence(context);

        assertEquals(5L, s.nextValue());
        assertEquals(6L, s.nextValue());

    }

    @TestTemplate
    void testS10(CallContext context) {
        S10Sequence s = new S10Sequence(context);

        assertEquals(-5L, s.nextValue());
        assertEquals(-4L, s.nextValue());

    }

    @TestTemplate
    void testS11(CallContext context) {
        S11Sequence s = new S11Sequence(context);

        assertEquals(1L, s.nextValue());
        assertEquals(2L, s.nextValue());

    }

    @TestTemplate
    void testS12(CallContext context) {
        S12Sequence s = new S12Sequence(context);

        assertEquals(1L, s.nextValue());
        assertEquals(2L, s.nextValue());
        assertEquals(3L, s.nextValue());
        assertEquals(4L, s.nextValue());
        assertEquals(5L, s.nextValue());
        //new cycle;
        assertEquals(1L, s.nextValue());

    }

    @TestTemplate
    void testS13(CallContext context) {
        S13Sequence s = new S13Sequence(context);

        assertEquals(1L, s.nextValue());
        assertEquals(0L, s.nextValue());
        assertEquals(-1L, s.nextValue());
        //new cycle;
        assertEquals(1L, s.nextValue());

    }

    @TestTemplate
    void testS14AndS15(CallContext context) {
        _testS14OrS15(new S14Sequence(context));
        _testS14OrS15(new S15Sequence(context));

    }

    @TestTemplate
    void testDefaultPkColumnValueWithSequence(CallContext context) {
        T1Cursor c = new T1Cursor(context);

        c.insert();
        assertEquals(4, c.getId().intValue());
        c.clear();

        c.insert();
        assertEquals(6, c.getId().intValue());


    }

    void _testS14OrS15(Sequence s) {
        long lastValue = s.nextValue();
        assertEquals(5L, lastValue);
        lastValue = s.nextValue();
        assertEquals(7L, lastValue);

        while (lastValue < 55) {
            lastValue = s.nextValue();
        }
        //new cycle;
        lastValue = s.nextValue();
        assertEquals(5L, lastValue);
    }
}
